jormungandr-watchdog/src/service/control.rs

```diff
@@ -0,0 +1,82 @@
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tokio::{
+    stream::Stream,
+    sync::watch::{self, Receiver, Sender},
+};
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum Control {
+    Shutdown,
+    Kill,
+}
+
+pub struct Controller {
+    sender: Sender<Control>,
+    receiver: Receiver<Control>,
+}
+
+pub struct ControlReader {
+    receiver: Receiver<Control>,
+}
+
+impl Controller {
+    #[allow(clippy::new_without_default)]
+    pub async fn new() -> Self {
+        let (sender, mut receiver) = watch::channel(Control::Kill);
+
+        let _ = receiver.recv().await;
+
+        Controller { sender, receiver }
+    }
+
+    pub fn reader(&self) -> ControlReader {
+        ControlReader {
+            receiver: self.receiver.clone(),
+        }
+    }
+
+    pub fn send(&mut self, control: Control) {
+        if self.sender.broadcast(control).is_err() {
+            // the `Controller` own a Receiver so broadcast
+            // cannot fail
+            unsafe { std::hint::unreachable_unchecked() }
+        }
+    }
+
+    pub async fn reset(&mut self) -> ControlReader {
+        let mut reader = self.reader();
+        self.send(Control::Kill);
+        if reader.updated().await.is_none() {
+            // `Controller` owns the sender and a send has just ben sent
+            // so it is impossible not to have an updated control
+            unsafe { std::hint::unreachable_unchecked() }
+        }
+        reader
+    }
+}
+
+impl ControlReader {
+    pub async fn updated(&mut self) -> Option<Control> {
+        self.receiver.recv().await
+    }
+}
+
+impl Future for ControlReader {
+    type Output = Option<Control>;
```

Comment:
What does `None` mean here? If it's not meant to be normal to drop the sender before sending the final command, perhaps make it a `Result` with an informative error?

Reply:
This should really be only a stream and not a `Future`. It is used as a `Stream` in the `service/mod.rs`'s `select!`. If you have a suggestion to use the `Stream` instead of the future here that would be welcome?

Reply:
I've removed the future and use `updated` instead. easier and cleaner